cardano-testnet/src/Testnet/Components/TestWatchdog.hs

```diff
@@ -0,0 +1,137 @@
+-- @
+-- runWithWatchdog watchdogConfig $ \watchdog -> do
+--   -- body of your test case
+-- @
+module Testnet.Components.TestWatchdog
+  ( runWithWatchdog_
+  , runWithWatchdog
+  , runWithDefaultWatchdog_
+  , runWithDefaultWatchdog
+  , Watchdog
+  , kickWatchdog
+  , poisonWatchdog
+  ) where
+
+import           Control.Concurrent (myThreadId, threadDelay, throwTo)
+import           Control.Concurrent.STM (atomically)
+import           Control.Concurrent.STM.TChan (TChan, newTChanIO, tryReadTChan, writeTChan)
+import           Control.Exception.Safe (Exception)
+import           Control.Monad.IO.Class
+import           Data.Time (NominalDiffTime, UTCTime, diffUTCTime, getCurrentTime,
+                   nominalDiffTimeToSeconds)
+import           GHC.Conc (ThreadId)
+import           GHC.Stack
+
+import qualified Hedgehog.Extras as H
+
+-- | Configuration for the watchdog.
+newtype WatchdogConfig = WatchdogConfig
+  { watchdogTimeout :: Int -- ^ Timeout in µs after which watchdog will kill the test case
+  }
+
+-- | Default watchdog config with 10 minutes timeout.
+defaultWatchdogConfig :: WatchdogConfig
+defaultWatchdogConfig = WatchdogConfig
+  { watchdogTimeout = 600_000_000
+  }
+
+-- | A watchdog
+data Watchdog = Watchdog
+  { watchdogConfig :: !WatchdogConfig
+  , watchedThreadId :: !ThreadId -- ^ monitored thread id
+  , startTime :: !UTCTime -- ^ watchdog creation time
+  , kickChan :: TChan WatchdogCommand -- ^ a queue of watchdog commands
+  }
+
+-- | Create a new watchdog
+makeWatchdog :: MonadIO m
+             => WatchdogConfig
+             -> ThreadId -- ^ thread id which will get killed after timeouts expire
+             -> m Watchdog
+makeWatchdog config watchedThreadId' = liftIO $ do
+  watchdog <- Watchdog config watchedThreadId' <$> getCurrentTime <*> newTChanIO
+  kickWatchdog watchdog
+  pure watchdog
+
+-- | Run watchdog in a loop
+runWatchdog :: MonadIO m
+            => Watchdog
+            -> m ()
+runWatchdog w@Watchdog{watchedThreadId, startTime, kickChan} = liftIO $ do
+  atomically (tryReadTChan kickChan) >>= \case
+    Just PoisonPill ->
+      -- deactivate watchdog
+      pure ()
+    Just (Kick timeout) -> do
+      -- got a kick, wait for another period
+      threadDelay timeout
+      runWatchdog w
+    Nothing -> do
+      -- we are out of scheduled timeouts, kill the monitored thread
+      currentTime <- getCurrentTime
+      throwTo watchedThreadId . WatchdogException $ diffUTCTime currentTime startTime
+
+-- | Watchdog command
+data WatchdogCommand
+  = Kick !Int -- ^ Add another delay
+  | PoisonPill -- ^ Stop the watchdog
+
+-- | Enqueue a kick for the watchdog. It will extend the timeout by another one defined in the watchdog
+-- configuration.
+kickWatchdog :: MonadIO m => Watchdog -> m ()
+kickWatchdog Watchdog{watchdogConfig=WatchdogConfig{watchdogTimeout}, kickChan} = liftIO $
+  atomically $ writeTChan kickChan (Kick watchdogTimeout)
```

Comment:
Need an example of `kickWatchdog` because it's difficult to compare this to alternate designs without understanding the use case.

Reply:
We don't need it for our codebase right now, but i's and escape hatch for watchdog. I see kicking as useful in a scenario when your test code branches out and you decide that you need more time to complete the operation but it still may hang.

Reply:
May I suggest computing the deadline and storing that in a `TVar` instead.  This would let you move the deadline further into the future to increase the amount of time for the test and also move the deadline towards the past to remove the amount of time as well.

Reply:
I think that would complicate things for a little gain. 
1. What's the use case for shortening the deadline? I think `byDeadlineM` is better suited for limiting the time spent on the computation.
1. How the deadline check should be realised? By busy-waiting and checking the expiration every second?

Reply:
I think if you only care about extending the timeout, a `TVar Int` is possibly still simpler.

```
kickWatchdog :: MonadIO m => Watchdog -> m ()
kickWatchdog Watchdog{watchdogConfig=WatchdogConfig{watchdogTimeout}, tTimeout} = liftIO $
  atomically $ modifyTVar tTimeout (+watchdogTimeout)
```

```
runWatchdog w@Watchdog{watchedThreadId, startTime, tTimeout} = liftIO $ do
  timeout <- atomically (swapTVar tTimeout 0) >>= \case
  
  if timeout >= 0
    then do
      -- got a kick, wait for another period
      threadDelay $ timeout * 1_000_000
      runWatchdog w
    else do
      -- we are out of scheduled timeouts, kill the monitored thread
      currentTime <- getCurrentTime
      throwTo watchedThreadId . WatchdogException $ diffUTCTime currentTime startTime
 ```